substrate/state-machine/src/overlayed_changes.rs

```diff
@@ -0,0 +1,326 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The overlayed changes to state.
+
+use std::collections::{HashMap, HashSet};
+use changes_trie::Configuration as ChangesTrieConfig;
+
+/// The overlayed changes to state to be queried on top of the backend.
+///
+/// A transaction shares all prospective changes within an inner overlay
+/// that can be cleared.
+#[derive(Debug, Default, Clone)]
+pub struct OverlayedChanges {
+	pub(crate) prospective: HashMap<Vec<u8>, (Option<Vec<u8>>, Option<HashSet<u32>>)>,
+	pub(crate) committed: HashMap<Vec<u8>, (Option<Vec<u8>>, Option<HashSet<u32>>)>,
+	pub(crate) changes_trie_config: Option<ChangesTrieConfig>,
+	pub(crate) block: Option<u64>,
+	pub(crate) extrinsic: Option<u32>,
+}
+
+impl OverlayedChanges {
+	/// Sets the changes trie configuration.
+	pub(crate) fn set_changes_trie_config(&mut self, block: u64, config: ChangesTrieConfig) {
```

Comment:
Setting extrinsic index is noop in case if config hasn't been set.
Setting different config leads to the creation of invalid changes trie.
Updated comment.